### PR TITLE
[VP][EVL] Support select instruction with EVL-vectorization

### PR DESCRIPTION
**Pull request by LiqinWeng**

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 1]
```diff
@@ -1082,6 +1100,36 @@ void VPWidenSelectRecipe::execute(VPTransformState &State) {
...
+  // loop. This means that we can't just use the original 'cond' value.
+  // We have to take the 'vectorized' value and pick the first lane.
+  // Instcombine will make this a no-op.
+  auto *InvarCond =
+      isInvariantCond() ? State.get(getCond(), VPIteration(0, 0)) : nullptr;
+
+  Value *Cond = InvarCond ? InvarCond : State.get(getCond(), 0);
+  if (!isa<VectorType>(Cond->getType())) {
+    Cond = BuilderIR.CreateVectorSplat(State.VF, Cond, "splat.cond");
+  }
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1122]: Drop braces here

File: llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-select-intrinsics.ll [hunk 2]
```diff
@@ -0,0 +1,101 @@
+; REQUIRES: asserts
```
Comment (alexey-bataev) [inline llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-select-intrinsics.ll:1]: Need to precommit a test in a separate patch

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -1082,6 +1100,36 @@ void VPWidenSelectRecipe::execute(VPTransformState &State) {
...
+  // The condition can be loop invariant but still defined inside the
+  // loop. This means that we can't just use the original 'cond' value.
+  // We have to take the 'vectorized' value and pick the first lane.
+  // Instcombine will make this a no-op.
+  auto *InvarCond =
+      isInvariantCond() ? State.get(getCond(), VPIteration(0, 0)) : nullptr;
+
+  Value *Cond = InvarCond ? InvarCond : State.get(getCond(), 0);
+  if (!isa<VectorType>(Cond->getType())) {
+    Cond = BuilderIR.CreateVectorSplat(State.VF, Cond, "splat.cond");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1121]: Do you have a test for this?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 1 above)
Comment (LiqinWeng) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1122]: Sorry, I didn't construct such a test case, but adding conditions may be safer

File: llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-select-intrinsics.ll (hunk 2 above)
Comment (LiqinWeng) [reply llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-select-intrinsics.ll:1]: spilt done

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (LiqinWeng) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1121]: Sorry, I didn't construct such a test case, but adding conditions may be safer

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1121]: Need a test

Comment (LiqinWeng) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Sorry, I tried to wirte such a test case, but it didn't sucess, so I will remove this condition for now.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 4]
```diff
@@ -1116,9 +1116,8 @@ void VPWidenSelectEVLRecipe::execute(VPTransformState &State) {
   auto *InvarCond =
       isInvariantCond() ? State.get(getCond(), VPIteration(0, 0)) : nullptr;
 
+  // FIXME: Do we have a scenario where cond is a scalar?
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1119]: Better to add assertion here

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1119]: I might be missing something, but can't this happen when the condition is a loop invariant value? If so, this should have a. test.

Comment (LiqinWeng) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1092]: added assertion

**Comment by LiqinWeng** [comment]
Any other questions? Can I merge it ?  @fhahn @alexey-bataev

**Comment by fhahn** [comment]
I am wondering if there's a better way to avoid duplicating all or most recipes for EVL. 

Would it be sufficient to create recipes to widen the appropriate VPIntrinsic directly at least for some cases? This would need allowing creating widened intrinsic recipes without underlying call instruction or function object (https://github.com/llvm/llvm-project/pull/110486). Then introducing vp.select (and others) can be added with a few lines like in https://github.com/llvm/llvm-project/pull/110489

**Comment by Mel-Chen** [comment]
> I am wondering if there's a better way to avoid duplicating all or most recipes for EVL.

I’ve also been seriously thinking about this recently. But my thinking is whether it’s really necessary to create an EVL version for each recipe, especially after noticing VLOpt https://github.com/llvm/llvm-project/pull/108640.

I think maybe we only need to create EVL recipes for a small subset of recipes, such as those related to memory accesses (load, store, interleaved accesses), in-loop reduction... etc. IIUC, VLOpt should have the ability to prune the VL.
There are two pros to this approach:
1. Minimizing the number of added recipes.
2. We can reuse optimizations, such as InstCombine. For example, if there is a combination rule for `add <vec1>, <vec2>`, we can share that rule without needing to add the similar comibination rule for `vp.add`.

In conclusion, we only need to transforms the partial recipes into EVL recipes. As for which ones exactly, we can determine by whether the recipe has a mask operand to decide if an EVL recipe is necessary.

What do you think?

**Comment by michaelmaitland** [comment]
> I’ve also been seriously thinking about this recently. But my thinking is whether it’s really necessary to create an EVL version for each recipe, especially after noticing VLOpt https://github.com/llvm/llvm-project/pull/108640.

If possible, I think it is better to avoid creating EVL that is larger than required. The first reason is that the VLOpt pass is not very mature and may miss some cases. The original intention of the pass is to fix the case introduced by vector GEP which has no EVL based version. In that case, there was no way for the compiler to use the correct VL at the IR level which makes that pass necessary. But if the compiler can avoid introducing larger than needed EVL, that is preferred, since it will create less work for the VLOpt pass, which will improve compiler times.

**Comment by alexey-bataev** [comment]
> > I’ve also been seriously thinking about this recently. But my thinking is whether it’s really necessary to create an EVL version for each recipe, especially after noticing VLOpt #108640.
> 
> If possible, I think it is better to avoid creating EVL that is larger than required. The first reason is that the VLOpt pass is not very mature and may miss some cases. The original intention of the pass is to fix the case introduced by vector GEP which has no EVL based version. In that case, there was no way for the compiler to use the correct VL at the IR level which makes that pass necessary. But if the compiler can avoid introducing larger than needed EVL, that is preferred, since it will create less work for the VLOpt pass, which will improve compiler times.

+1

**Comment by LiqinWeng** [comment]
> In conclusion, we only need to transforms the partial recipes into EVL recipes. As for which ones exactly, we can determine by whether the recipe has a mask operand to decide if an EVL recipe is necessary.

If triple count is not a constant, can it be directly introduced into EVL?
